### PR TITLE
Enable fips mode when building envoy

### PR DESCRIPTION
**Pull request by bathina2**
## Description
The following PR enables building envoy/emissary in FIPS mode. This will ensure that the envoy uses FIPS certified cryptographic algorithms when decrypting data.
Developers can set this by using the `export FIPS_MODE=true` option before doing `make update-base`.

## Related Issues
https://github.com/emissary-ingress/emissary/issues/4277

## Testing
After adding my changes I was able to to build envoy using `make update-base`. The resulting image was tagged as a FIPS enabled image-

```
# docker run -it --entrypoint envoy-static-stripped  emissaryingress/base-envoy:envoy-0.-.opt --version

envoy-static-stripped  version: 4ce93dc3ace00ae9108b179d0afaceac13f4602a/1.17.4/Modified/RELEASE/BoringSSL-FIPS
```

## Checklist

 - [x] I made sure to update `CHANGELOG.md`. _Not required_

   Remember, the CHANGELOG needs to mention:
    + Any new features
    + Any changes to our included version of Envoy
    + Any non-backward-compatible changes
    + Any deprecations

 - [x] This is unlikely to impact how Ambassador performs at scale.

   Remember, things that might have an impact at scale include:
    + Any significant changes in memory use that might require adjusting the memory limits
    + Any significant changes in CPU use that might require adjusting the CPU limits
    + Anything that might change how many replicas users should use
    + Changes that impact data-plane latency/scalability

 - [x] My change is adequately tested.
    Not sure exactly how I can Unit test my changes. I have automations in place that follow the described build steps to create an Envoy image. These images do contain the FIPS suffix at the end of the version. 
    When omitting the option, it generates the appropriate image with out the suffix.

   Remember when considering testing:
    + Your change needs to be specifically covered by tests.
       + Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
    + You also need to make sure that the _entire area being changed_ has adequate test coverage.
       + If existing tests don't actually cover the entire area being changed, add tests.
       + This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
    + We should lean on the bulk of code being covered by unit tests, but...
    + ... an end-to-end test should cover the integration points

 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

 - [x] The changes in this PR have been reviewed for security concerns and adherence to security best practices.

### CODE REVIEW

File: _cxx/envoy.mk [hunk 1]
```diff
@@ -7,6 +7,8 @@ YES_I_AM_OK_WITH_COMPILING_ENVOY ?=
 ENVOY_TEST_LABEL ?= //test/...
 # Set RSYNC_EXTRAS=Pv or something to increase verbosity.
 RSYNC_EXTRAS ?=
+# Set to non-empty to enable compiling Envoy in FIPS mode.
+FIPS_MODE ?=
```
Comment (LukeShu) [inline _cxx/envoy.mk:11]: If it affects the Envoy build, then it's gotta be included in the image tag and in the section of variables that `IF YOU MESS WITH ANY OF THESE VALUES, YOU MUST RUN make update-base.`

File: _cxx/envoy.mk [hunk 2]
```diff
@@ -206,7 +208,7 @@ $(OSS_HOME)/docker/base-envoy/envoy-static: $(ENVOY_BASH.deps) FORCE
 	            exit 1; \
 	        fi; \
 	        $(call ENVOY_BASH.cmd, \
-	            $(ENVOY_DOCKER_EXEC) bazel build --verbose_failures -c $(ENVOY_COMPILATION_MODE) --config=clang //source/exe:envoy-static; \
+	            $(ENVOY_DOCKER_EXEC) bazel build $$(if [ -n '$(FIPS_MODE)' ]; then echo "--define boringssl=fips"; fi) --verbose_failures -c $(ENVOY_COMPILATION_MODE) --config=clang //source/exe:envoy-static; \
```
Comment (LukeShu) [inline _cxx/envoy.mk:211]: This can be better written with Make macros: `$(if $(FIPS_MODE), --define boringssl=fips )`.

File: _cxx/envoy.mk (hunk 1 above)
Comment (bathina2) [reply _cxx/envoy.mk]: Updated

File: _cxx/envoy.mk (hunk 2 above)
Comment (bathina2) [reply _cxx/envoy.mk]: Thanks for the feedback, not super familiar with using make macros, but tested with the change and it did infact set it correctly!

**Comment by bathina2** [comment]
@LukeShu I updated `DEVELOPING.md` with better messaging around what is supported. Also removed the release note.

File: _cxx/envoy.mk [hunk 3]
```diff
@@ -23,6 +23,9 @@ RSYNC_EXTRAS ?=
   ENVOY_DOCKER_VERSION ?= $(BASE_ENVOY_RELVER).$(ENVOY_COMMIT).$(ENVOY_COMPILATION_MODE)
   ENVOY_DOCKER_TAG ?= $(ENVOY_DOCKER_REPO):envoy-$(ENVOY_DOCKER_VERSION)
   ENVOY_FULL_DOCKER_TAG ?= $(ENVOY_DOCKER_REPO):envoy-full-$(ENVOY_DOCKER_VERSION)
+
+  # Set to non-empty to enable compiling Envoy in FIPS mode.
+  FIPS_MODE ?=
```
Comment (kflynn) [inline _cxx/envoy.mk:28]: I'm pretty sure that you also must include `$(FIPS_MODE)` in `ENVOY_DOCKER_VERSION` for things to work correctly:

```
ENVOY_DOCKER_VERSION ?= $(BASE_ENVOY_RELVER).$(ENVOY_COMMIT).$(ENVOY_COMPILATION_MODE).$(if $(FIPS_MODE),FIPS,noFIPS)
```

or the like. I'll defer to @LukeShu here though.

**Review by kflynn — CHANGES_REQUESTED** [review]
I think you're going to have to update `ENVOY_DOCKER_VERSION` too, or things won't work.

File: _cxx/envoy.mk (hunk 3 above)
Comment (bathina2) [reply _cxx/envoy.mk]: Thanks for the suggestion @kflynn!

**Comment by ddymko** [comment]
@bathina2 @kflynn 

I had made a change in https://github.com/emissary-ingress/emissary/pull/4377/commits/38f9fd62d324efddb151e9d61b6e6000d1df25c4 in order for CI to pass.

I noticed that if `.$(if $(FIPS_MODE),FIPS,noFIPS)`  was set we would be looking for builds with `.noFIPS` which we don't have. I modified this to `$(if $(FIPS_MODE),.FIPS)` to prevent the the CI failures. 

@bathina2 Could you please pull in this change and also revise your commit history to a single commit?

**Comment by bathina2** [comment]
@ddymko Thanks for the review and suggestions. I added your changes and squashed everything into a single commit.

**Review by ddymko — APPROVED** [review]
Changes look good - CI passed in https://github.com/emissary-ingress/emissary/pull/4377

**Pull request merged**